File: backend_service/inference/binaries.py
"""Binary + Python-runtime resolution for the inference layer.

Where to find ``llama-server`` / ``llama-server-turbo`` / ``llama-cli``,
which Python interpreter to use for the MLX worker, and a small JSON
subprocess wrapper used by the capability probe and other helpers.

Extracted from ``inference/__init__.py`` as part of the v0.8.0 refactor.
"""

from __future__ import annotations

import json
import os
import shutil
import subprocess
import sys
from pathlib import Path
from typing import Any

from backend_service.inference._constants import WORKSPACE_ROOT


def _json_subprocess(
    command: list[str],
    *,
    timeout: float = 15.0,
    cwd: Path = WORKSPACE_ROOT,
) -> tuple[int, dict[str, Any] | None, str]:
    try:
        completed = subprocess.run(
            command,
            cwd=str(cwd),
            check=False,
            capture_output=True,
            timeout=timeout,
        )
    except (OSError, subprocess.TimeoutExpired) as exc:
        return (-1, None, str(exc))

    payload: dict[str, Any] | None = None
    stdout = completed.stdout.decode("utf-8", errors="replace").strip()
    stderr = completed.stderr.decode("utf-8", errors="replace").strip()
    if stdout:
        try:
            payload = json.loads(stdout)
        except json.JSONDecodeError:
            payload = None
    return (completed.returncode, payload, stderr or stdout)


def _resolve_mlx_python() -> str:
    override = os.getenv("CHAOSENGINE_MLX_PYTHON")
    if override:
        return override
    candidate = WORKSPACE_ROOT / ".venv" / "bin" / "python"
    if candidate.exists():
        return str(candidate)
    return sys.executable


# Common install locations for llama.cpp binaries that may not be in PATH
# when launched from a GUI app (Tauri doesn't inherit the user's shell profile).
_CHAOSENGINE_BIN_DIR = str(Path.home() / ".chaosengine" / "bin")

_LLAMA_FALLBACK_DIRS = [
    _CHAOSENGINE_BIN_DIR,        # ChaosEngineAI-managed binaries
    "/opt/homebrew/bin",         # macOS ARM Homebrew
    "/usr/local/bin",            # macOS Intel Homebrew / manual
    "/usr/bin",                  # system
    str(Path.home() / ".local" / "bin"),  # pip --user installs
]


def _which_with_fallbacks(name: str) -> str | None:
    """Like shutil.which but also checks common install directories."""
    found = shutil.which(name)
    if found:
        return found
    for d in _LLAMA_FALLBACK_DIRS:
        candidate = os.path.join(d, name)
        if os.path.isfile(candidate) and os.access(candidate, os.X_OK):
            return candidate
    return None


def _resolve_llama_server() -> str | None:
    override = os.getenv("CHAOSENGINE_LLAMA_SERVER")
    if override:
        return override
    return _which_with_fallbacks("llama-server")


def _resolve_llama_server_turbo() -> str | None:
    """Resolve the TurboQuant fork of llama-server (``llama-server-turbo``).

    This fork supports all standard cache types **plus** iso/planar/turbo
    cache types required by RotorQuant and TurboQuant strategies.
    """
    override = os.getenv("CHAOSENGINE_LLAMA_SERVER_TURBO")
    if override:
        return override
    return _which_with_fallbacks("llama-server-turbo")


def _resolve_llama_cli() -> str | None:
    override = os.getenv("CHAOSENGINE_LLAMA_CLI")
    if override:
        return override
    return _which_with_fallbacks("llama-cli")

File: src-tauri/src/binaries.rs
//! Resolves paths to bundled binaries (llama-server, sd, etc.).
//!
//! Each `resolve_*` honours an env-var override first, falls back to
//! `~/.chaosengine/bin/<name>` for the standard / turbo / sd-cli
//! managed installs, then finally walks `PATH`. The two utility
//! helpers (`resolve_candidate`, `find_in_path`) are shared with
//! `lib.rs` for cases that don't need the per-binary fallback chain.
//!
//! Extracted from `src-tauri/src/lib.rs` as part of the v0.8.0
//! Phase 3 refactor.

use std::env;
use std::path::{Path, PathBuf};


pub fn resolve_llama_server(_workspace_root: &Path) -> Option<PathBuf> {
    if let Some(value) = env::var_os("CHAOSENGINE_LLAMA_SERVER") {
        if let Some(path) = resolve_candidate(value) {
            return Some(path);
        }
    }

    find_in_path(&["llama-server"])
}

pub fn resolve_llama_server_turbo(_workspace_root: &Path) -> Option<PathBuf> {
    if let Some(value) = env::var_os("CHAOSENGINE_LLAMA_SERVER_TURBO") {
        if let Some(path) = resolve_candidate(value) {
            return Some(path);
        }
    }

    // Check ~/.chaosengine/bin/ first (ChaosEngineAI-managed installs),
    // then fall back to PATH.
    if let Ok(home) = env::var("HOME") {
        let managed = PathBuf::from(home)
            .join(".chaosengine")
            .join("bin")
            .join("llama-server-turbo");
        if managed.exists() {
            return Some(managed);
        }
    }

    find_in_path(&["llama-server-turbo"])
}

pub fn resolve_llama_cli(_workspace_root: &Path) -> Option<PathBuf> {
    if let Some(value) = env::var_os("CHAOSENGINE_LLAMA_CLI") {
        if let Some(path) = resolve_candidate(value) {
            return Some(path);
        }
    }

    find_in_path(&["llama-cli"])
}

pub fn resolve_sd_cpp(_workspace_root: &Path) -> Option<PathBuf> {
    if let Some(value) = env::var_os("CHAOSENGINE_SDCPP_BIN") {
        if let Some(path) = resolve_candidate(value) {
            return Some(path);
        }
    }

    if let Ok(home) = env::var("HOME") {
        let managed = PathBuf::from(home).join(".chaosengine").join("bin").join("sd");
        if managed.exists() {
            return Some(managed);
        }
    }

    find_in_path(&["sd"])
}

pub fn resolve_candidate(value: impl Into<PathBuf>) -> Option<PathBuf> {
    let candidate = value.into();
    if candidate.exists() {
        return Some(candidate);
    }
    if candidate.components().count() == 1 {
        return find_in_path(&[candidate.to_string_lossy().as_ref()]);
    }
    None
}

pub fn find_in_path(names: &[&str]) -> Option<PathBuf> {
    let path_var = env::var_os("PATH")?;
    for directory in env::split_paths(&path_var) {
        for name in names {
            let candidate = directory.join(name);
            if candidate.exists() {
                return Some(candidate);
            }
            #[cfg(windows)]
            {
                let exe_candidate = directory.join(format!("{name}.exe"));
                if exe_candidate.exists() {
                    return Some(exe_candidate);
                }
            }
        }
    }
    None
}
